### PR TITLE
[AMDGPU] Add patterns for V_CMP_O/U

### PR DESCRIPTION
**Pull request by Pierre-vh**
Fixes SWDEV-427162

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/VOPCInstructions.td [hunk 1]
```diff
@@ -1110,6 +1114,8 @@ defm : FCMP_Pattern <COND_ULT, V_CMP_NGE_F64_e64, f64>;
 defm : FCMP_Pattern <COND_ULE, V_CMP_NGT_F64_e64, f64>;
 
 let OtherPredicates = [HasTrue16BitInsts] in {
+defm : FCMP_Pattern <COND_O, V_CMP_U_F16_t16_e64, f16>;
+defm : FCMP_Pattern <COND_UO, V_CMP_O_F16_t16_e64, f16>;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/VOPCInstructions.td:1118]: These two look backwards.

File: llvm/lib/Target/AMDGPU/VOPCInstructions.td [hunk 2]
```diff
@@ -1126,6 +1132,8 @@ defm : FCMP_Pattern <COND_ULE, V_CMP_NGT_F16_t16_e64, f16>;
 } // End OtherPredicates = [HasTrue16BitInsts]
 
 let OtherPredicates = [NotHasTrue16BitInsts] in {
+defm : FCMP_Pattern <COND_O, V_CMP_U_F16_e64, f16>;
+defm : FCMP_Pattern <COND_UO, V_CMP_O_F16_e64, f16>;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/VOPCInstructions.td:1136]: These two look backwards.

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fcmp.w32.ll [hunk 3]
```diff
@@ -494,6 +494,121 @@ define amdgpu_kernel void @v_fcmp_f32_ole(ptr addrspace(1) %out, float %src) {
...
+; GISEL-GFX10-NEXT:    s_clause 0x1
+; GISEL-GFX10-NEXT:    s_load_dword s4, s[0:1], 0x2c
+; GISEL-GFX10-NEXT:    s_load_dwordx2 s[2:3], s[0:1], 0x24
+; GISEL-GFX10-NEXT:    v_mov_b32_e32 v1, 0
+; GISEL-GFX10-NEXT:    s_waitcnt lgkmcnt(0)
+; GISEL-GFX10-NEXT:    v_cmp_u_f32_e64 s0, 0x42c80000, s4
+; GISEL-GFX10-NEXT:    v_mov_b32_e32 v0, s0
+; GISEL-GFX10-NEXT:    global_store_dword v1, v0, s[2:3]
+; GISEL-GFX10-NEXT:    s_endpgm
+  %result = call i32 @llvm.amdgcn.fcmp.f32(float %src, float 100.00, i32 8)
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fcmp.w32.ll:608]: Can the problem also occur when you use regular fcmp instructions (possibly combined with a ballot intrinsic)? Or does it only affect the llvm.amdgcn.fcmp intrinsic?

The llvm.amdgcn.fcmp intrinsic should be deprecated.

Comment (Pierre-vh) [reply llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fcmp.w32.ll:608]: I'm not sure really, I observed this due to a fcmp being generated from InstCombineIntrinsics by combining a icmp + fcmp
In this case the `setcc setuo` is emitted from FCMP lowering. Maybe the actual fcmp inst goes through a different path and the o/uo gets legalized away?

**Review by jayfoad — APPROVED** [review]

**Pull request merged**